Allow extra searchable metadata in package

mobileoverlord opened the pull request:
WIP

Added fields and tests to have package_metadata contain extra map.

Todo:
extend search API to pass object path and search words

Tracked alongside
https://github.com/hexpm/hex/pull/234
https://github.com/hexpm/specifications/pull/6

mobileoverlord commented:
How should we handle bringing in the ability to match on extra path

Lets say we have several packages with extra package metadata defined like this:
`%{"extra" => %{"foo" => "bar"}}`

How to search for nerves packages where the key foo in extra = bar
https://hex.pm/packages?_utf8=%E2%9C%93&search=nerves&sort=downloads&extra:foo=bar

ericmj commented:
Something like this maybe? https://hex.pm/packages?_utf8=%E2%9C%93&search=nerves&sort=downloads&extra_path[]=foo&extra_path[]=bar&extra_value=baz

We need to be able to support nested paths, I think.

ericmj commented on priv/repo/migrations/20160531152824_add_package_extra_index.exs:
```diff
@@ -0,0 +1,14 @@
+    execute """
+      CREATE INDEX packages_extra_text ON
+        packages USING GIN (to_tsvector('english', regexp_replace((meta->'extra')::text, '/', ' ')))
+    """
```
This will not work if you want to specify paths so we can't use full text search unfortunately. This code is encoding the json as text so you will have a string with "{...}".

If we wan't full text search we need to concatenate all text inside the nested json values. I think exact value search would be better.

ericmj commented on test/models/package_test.exs:
```diff
@@ -51,4 +51,26 @@ defmodule HexWeb.PackageTest do
+    user = HexWeb.Repo.get_by!(User, username: "eric")
+    assert {:ok, package} =
+      Package.build(user, pkg_meta(%{name: "nerves", description: "DSL"}))
+      |> HexWeb.Repo.insert
```
Use bang version of `insert`.

ericmj commented on test/models/package_test.exs:
```diff
@@ -51,4 +51,26 @@ defmodule HexWeb.PackageTest do
+
+    assert {:ok, package} =
+      Package.update(package, %{"meta" => meta})
+      |> HexWeb.Repo.update
```
Use bang version of `update`.

ericmj commented on web/models/package_metadata.ex:
```diff
@@ -6,10 +6,11 @@ defmodule HexWeb.PackageMetadata do
     field :licenses, {:array, :string}
     field :links, :map
     field :maintainers, {:array, :string}
+    field :extra, :map
```
We need to add this field to audit_log.ex also. https://github.com/hexpm/hex_web/blob/master/web/models/audit_log.ex#L38

wojtekmach replied:
btw, when we update to latest Ecto, this map could be typed. It goes back to @mobileoverlord's question if this should be map of maps, or map of arrays (ideally not both, so that we can get type safety)

mobileoverlord commented:
I started putting in some test information.
`{"id": "12345", "extra": {"foo": {"bar": "baz"}, "key": "value", "list": ["a", "b", "c"]}, "links": {"Github": "http://example.com/github"}, "licenses": ["Apache 2.0"], "description": "description", "maintainers": ["Justin Schneck", "Frank Hunleth"]}`

Should we only support nested paths, or can we also support nested lists

mobileoverlord commented:
nested paths are pretty simple
```
SELECT p0 FROM "packages" AS p0 
WHERE p0.meta #>> '{extra, key}'= 'value'
```
I was thinking we could split the valiue of the extra path param with `:`
?extra_path=foo:bar&extra_value=baz

ericmj commented:
I've been looking more at JSON value indexing at https://www.postgresql.org/docs/9.5/static/datatype-json.html. It seems we can only query using the index on top level values, but we can query on array elements. If we want to query on nested elements we need to flatten the JSON object inside the index expression.

> I was thinking we could split the valiue of the extra path param with :
?extra_path=foo:bar&extra_value=baz

This assumes that the keys cannot contain `:`. That's fair but would need to be documented in the specification and rejected in validation.

> do we or can we support nested lists. In the example above if trying to select form a list
?extra_path=list&extra_value=a
could return results where a in list

We can support that on the first "level" of a list but nesting the path inside lists would be tricky so I don't think we should allow that.

ericmj commented:
We can query on one level of nesting if the nested element is an array as shown in the example from the documentation:

```
{
    "guid": "9c36adc1-7fb5-4d5b-83b4-90356a46061a",
    "name": "Angela Barton",
    "is_active": true,
    "company": "Magnafone",
    "address": "178 Howard Place, Gulf, Washington, 702",
    "registered": "2009-11-07T08:53:22 +08:00",
    "latitude": 19.793713,
    "longitude": 86.513373,
    "tags": [
        "enim",
        "aliquip",
        "qui"
    ]
}

-- Find documents in which the key "tags" contains key or array element "qui"
SELECT jdoc->'guid', jdoc->'name' FROM api WHERE jdoc -> 'tags' ? 'qui';
```

mobileoverlord commented:
That syntax works really well for both the value being a list or a value. I think that the strategy is:

nested maps
%{"foo" => %{"bar" => "baz"}}

nested lists
%{"foo" => ["bar", "baz"]}

**not** maps in lists
%{"foo" => [%{"bar" => "baz"}]}

As per the params

Is it safe to enum params and find the one with the key that starts_with? "extra:" and then break the path off the remainder of the key. This would let us do
`?extra:foo:bar=baz`

ericmj commented:
> nested maps

Keep in mind that this would require us to flatten json object in the index expression (if I understood the postgres docs correctly).

> Is it safe to enum params and find the one with the key that starts_with? "extra:" and then break the path off the remainder of the key. This would let us do ?extra:foo:bar=baz

It would be possible, but with `?extra_path[]=...` that would not be needed since phoenix/plug will convert that to a list for us.

mobileoverlord commented:
Im getting caught up in trying to compose search in terms of a text fragments in to a query. After a talk with Eric, we determined that it would make more sense to incorporate searching on extra into the standard search behavior. For example

**Current Search**
`nerves_pkg`
searches:
`name OR description`

**New Search**
searching extra metadata values is exact and therefore should be joined with AND. The proposition is to allow expression in the search value. 

`name:nerves_pkg OR description:\"nerves_pkg\"`

backwards compatibility is achieved by converting a string with no keywords ("name", "description", "extra") into the above example.

The problems start to arise when introducing extra

`name:nerves_pkg OR description:\"nerves_pkg\" AND extra:"{\"a\" => \"b\"}"`

When parsed, the AND takes precedence so description AND extra... OR name. To get around this we would have to introduce `()`

`(name:nerves_pkg OR description:\"nerves_pkg\") AND extra:"{\"a\" => \"b\"}"`

This is where things start to feel like they are getting out of hand. Composing `where/3` concats with AND not OR, also, how do we dynamically add the `()` for grouping?

My main use case for this feature is to allow filtering search results so we can list all packages where name like 'nerves_pkg_%' and extra:{"nerves": {"binaries": [...]}} contains "this_binary". Offering _everything_ through an expressive search is feeling daunting. Wondering if anyone has an recommendations or different angles on the best implementation for everyone.

mobileoverlord commented:
for updated notes, we also ran experiments on how to execute using a single operand on jsonb
```
SELECT '{"a": {"b":[1,2], "d": 1}, "c": [1, 2]}'::jsonb @> '{"a": {"d": 1}}'::jsonb;
SELECT '{"a": {"b":[1,2], "d": 1}, "c": [1, 2]}'::jsonb @> '{"a": {"b": [2]}}'::jsonb;
```

ericmj commented:
You can make it always AND as long as we keep the current behaviour of ORing name and description when doing the "standard" search.

mobileoverlord commented:
Ready for another pass

mobileoverlord commented:
I changed the search semantics a little for name:v to expect the wildcards to be passed in as part of the name:nerves%

rebased and resolved conflicts

ericmj commented on web/controllers/api/package_controller.ex:
```diff
@@ -6,7 +6,7 @@ defmodule HexWeb.API.PackageController do
   def index(conn, params) do
     page     = HexWeb.Utils.safe_int(params["page"])
-    search   = HexWeb.Utils.safe_search(params["search"])
+    search   = params["search"]
```
We still need the conversion to `nil` for `""`. Maybe add a new function to Package that only does the String.replace?

ericmj commented on test/models/package_test.exs:
```diff
@@ -51,4 +51,37 @@ defmodule HexWeb.PackageTest do
+    search = [
+      {"name:nerves extra:list:[a]", 1},
+      {"name:nerves% extra:foo:bar:baz", 2},
+      {"name:nerves% extra:list:[1]", 1}]
```
I would rather separate by comma, rather than reusing `:`.

"name:nerves% extra:foo,bar,baz"

ericmj commented on web/models/package.ex:
```diff
@@ -115,26 +116,77 @@ defmodule HexWeb.Package do
-    search(query, filter)
+    if Enum.any?(@search_keys, & String.starts_with?(search, &1)) do
+      parse_search(search)
+      |> Enum.reduce(query, fn({k, v}, q) ->
```
Don't pipe here, instead: `Enum.reduce(parse_search(search), query, fn({k, v}, q) ->`.

ericmj commented on web/models/package.ex:
```diff
@@ -115,26 +116,77 @@ defmodule HexWeb.Package do
+    end)
+  end
+  defp extra_value(v),
+    do: Utils.safe_search(v)
```
Do we need to support list values?

mobileoverlord replied:
when the value to test is in a list it needs to be passed as such so we know to convert it to an actual list.

ericmj replied:
Do you have a need for it in nerves? Otherwise I would like to keep the implementation simple. If you do have a use case for it we should of course keep it.

ericmj commented on web/models/package.ex:
```diff
@@ -115,26 +116,77 @@ defmodule HexWeb.Package do
+    [h | t] = p
+
+    value =
+      extra_value(v)
```
Make this one line.

ericmj commented on web/models/package.ex:
```diff
@@ -115,26 +116,77 @@ defmodule HexWeb.Package do
+      extra_value(v)
+
+    extra =
+      extra(t, Map.put(%{}, h, value))
```
Make this one line as well.

ericmj commented on web/models/package.ex:
```diff
@@ -115,26 +116,77 @@ defmodule HexWeb.Package do
+    [v | p] =
+    search
+    |> String.split(":")
+    |> Enum.reverse
```
Indent after `=`.

ericmj commented on web/models/package.ex:
```diff
@@ -169,4 +221,43 @@ defmodule HexWeb.Package do
+  defp parse_key(string, search) do
+    {k, v} =
+      Regex.split(~r/:/, string, parts: 2)
+      |> List.to_tuple
```
Use `String.split` instead. No need to convert to a tuple.

ericmj commented on web/models/package.ex:
```diff
@@ -169,4 +221,43 @@ defmodule HexWeb.Package do
   end
+
+  defp parse_search(_, _ \\ [])
+  defp parse_search("", [nil | keys]), do: Enum.reverse(keys)
```
How can we get `nil` here?

ericmj commented on web/models/package.ex:
```diff
@@ -169,4 +221,43 @@ defmodule HexWeb.Package do
+      Regex.split(~r/:/, string, parts: 2)
+      |> List.to_tuple
+    {v, tail} =
+      parse_value(v)
```
Single line :)

ericmj commented on web/models/package.ex:
```diff
@@ -169,4 +221,43 @@ defmodule HexWeb.Package do
+      |> List.to_tuple
+    {v, tail} =
+      parse_value(v)
+    {add_param({String.to_atom(k), v}, search), tail}
```
`String.to_existing_atom`.

ericmj commented on web/models/package.ex:
```diff
@@ -169,4 +221,43 @@ defmodule HexWeb.Package do
+      [value] -> {value, ""}
+      [value, tail] -> {value, tail}
+    end
+  end
```
I would rewrite this function to something like this:

```elixir
defp parse_value(input) do
  {rest, delim} = case input do
    "\"" <> rest -> {rest, delim}
    _ -> {input, " "}
  end

  case String.split(rest, delim, parts: 2) do
    [value] -> {value, ""}
    [value, tail] -> {value, tail}
  end
end
```

It's hard to follow the follow the flow when the `nil` parameter is used as state.

ericmj replied:
You could also write it like you did, but remove the second argument and do the `String.split` in a `do_parse_value` function.

ericmj commented on web/models/package.ex:
```diff
@@ -169,4 +221,43 @@ defmodule HexWeb.Package do
+  defp add_param(param, [nil | tail]),
+    do: [param | tail]
+  defp add_param(param, [scope | tail]),
+    do: [param | [scope | tail]]
```
I don't understand this code. Why do we need it? It seems to me `parse_value` never returns lists so when can `scope` be a list? And from where do we get the `nil`?

mobileoverlord commented:
round 3

ericmj commented:
I think it looks good. Do we need to add an index to extra also?

mobileoverlord commented:
I was going to ask you that. I didn't know

ericmj commented:
I do think we need to add the index as a migration. Let me know if you want me to add it.

mobileoverlord commented:
I added the index.

ericmj commented on priv/repo/migrations/20160610143806_add_meta_extra_index.exs:
```diff
@@ -0,0 +1,14 @@
+  def up do
+    execute """
+      CREATE INDEX packages_meta_extra ON
+        packages USING GIN ((meta->'extra'))
```
We should use the index operator `jsonb_path_ops`, i.e. `USING GIN (meta->'extra' jsonb_path_ops)`. While you're at it append `_idx` to the index name :).

This is my final comment I believe, will merge after this change.

ericmj commented:
:heart:
